### ISSUE TITLE
Loading the weights from old model and continuing training apparently broken

### ISSUE THREAD

**Issue by mfernezir** [NVIDIA/DIGITS#533]
Scenario:

Model trained for 100 epochs and got to 50% accuracy with validation graphs looking good (yay). Loaded the model weights for another go with nothing changed (cloned job and loaded weights): flat line under 10 percent for 22 epochs. I've tried continuing many other jobs and it's always similar. Model weights apparently get loaded, but results are nowhere near where they were after the first run.

(3.1. dev)

**Comment by gheinrich** [comment]
Hi tried this on the tip of the `master` branch using Alexnet and didn't notice exactly the same thing. I trained Alexnet for 10 epochs on a 10-class dataset:
![alexnet-initial](https://cloud.githubusercontent.com/assets/3889770/12481539/35a325e2-c04a-11e5-969c-9f0d2e8cbffc.png)

Then I trained another instance of Alexnet selecting the 10 epoch of the previous model:
![alexnet-retrained](https://cloud.githubusercontent.com/assets/3889770/12481469/b92281d4-c049-11e5-9558-893eeffb27a7.png)

Except for the artefact (?) on iteration `0`, the model seems to start around where the original model left off (admittedly, the accuracy is about 10% lower, which sounds a little strange but perhaps this is not unexpected given that the learning rate is quite a bit higher on first iterations so the risk of overfitting the training set is higher).

Is this what you're seeing? From your description it isn't clear whether 10% accuracy means that the model hasn't learned anything (i.e. do you have many more than 10 classes?).

**Comment by gheinrich** [comment]
Actually the low accuracy on iteration 0 is not an artefact and the lower accuracy on first epoch after re-training is not due to overfit!

Part of the problem is the fact that the very last `InnerProduct` layer in the network is renamed to `xxx_retrained` so Caffe thinks it's a new layer and we have to totally retrain this layer from scratch. If I comment out these [lines](https://github.com/NVIDIA/DIGITS/blob/master/digits/frameworks/caffe_framework.py#L97-L99) then the problem disappears. I thought the renaming of the last layer was to prevent an error when retraining a new network with a different number of outputs but when I tried to train on a 2-class dataset after training on a 10-class dataset this worked seamlessly. 

@lukeyeager any comment?

**Comment by lukeyeager** [comment]
> I thought the renaming of the last layer was to prevent an error when retraining a new network with a different number of outputs but when I tried to train on a 2-class dataset after training on a 10-class dataset this worked seamlessly.

You're right - that's why we do it. When I comment out those lines and try to retrain on a dataset with a different number of labels, I get error messages like this:
```
2016-01-21 08:58:36 [20160121-085831-93e4] [ERROR] Train Caffe Model: Cannot copy param 0 weights from layer 'fc8'; shape mismatch.  Source param shape is 2 4096 (8192); target param shape is 20 4096 (81920). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.
2016-01-21 08:58:44 [20160121-085838-0a15] [ERROR] Train Caffe Model: Cannot copy param 0 weights from layer 'fc8'; shape mismatch.  Source param shape is 20 4096 (81920); target param shape is 2 4096 (8192). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.

```

**Comment by gheinrich** [comment]
Shall we rename the last `InnerProduct` layer only if the number of outputs is different? Or perhaps it would be more accurate to rename the layer only if the dataset is different. When using the same dataset we'd be able to re-use what the previous model learned about that last layer.

**Comment by lukeyeager** [comment]
If we're talking about running the same model with the same weights on the same dataset, then that sounds more like "continued training" or something other than "retraining". Maybe that should be a different feature?

**Comment by gheinrich** [comment]
I am not sure what the best term is for this. What I mean is if you're re-training a model on the same dataset you might expect the new model to immediately exhibit the same accuracy as the original model. From my understanding of this bug report this is what @mfernezir is expecting too. In order to achieve this I think we need to re-use the weights of the last layer too and I reckon this implies that we shouldn't rename the layer.

**Comment by lukeyeager** [comment]
You're right - we could add a check for that. I had gone with the conservative choice initially, but we could be a bit smarter.

**Comment by mfernezir** [comment]
Thanks for looking into this. It was a big surprise when validation accuracy dropped from learned 50% to only 7% and with no movement to previous numbers anytime soon.
